tests/Avalonia.UnitTests/TouchTestHelper.cs

```diff
@@ -27,18 +27,30 @@ public void Down(Interactive target, Interactive source, Point position = defaul
 
         public void Move(Interactive target, Interactive source, in Point position, KeyModifiers modifiers = default)
         {
-            target.RaiseEvent(new PointerEventArgs(InputElement.PointerMovedEvent, source, _pointer, (Visual)target, position,
-                Timestamp(), PointerPointProperties.None, modifiers));
+            var e = new PointerEventArgs(InputElement.PointerMovedEvent, source, _pointer, (Visual)target, position,
+                Timestamp(), PointerPointProperties.None, modifiers);
+            if (_pointer.CapturedGestureRecognizer != null)
+                _pointer.CapturedGestureRecognizer.PointerMoved(e);
+            else
+                target.RaiseEvent(e);
+
         }
 
         public void Up(Interactive target, Point position = default, KeyModifiers modifiers = default)
             => Up(target, target, position, modifiers);
 
         public void Up(Interactive target, Interactive source, Point position = default, KeyModifiers modifiers = default)
         {
-            source.RaiseEvent(new PointerReleasedEventArgs(source, _pointer, (Visual)target, position, Timestamp(), PointerPointProperties.None,
-                modifiers, MouseButton.None));
+            var e = new PointerReleasedEventArgs(source, _pointer, (Visual)target, position, Timestamp(), PointerPointProperties.None,
+                modifiers, MouseButton.None);
+
+            if (_pointer.CapturedGestureRecognizer != null)
+                _pointer.CapturedGestureRecognizer.PointerReleased(e);
+            else
+                source.RaiseEvent(e);
+
             _pointer.Capture(null);
```

Comment:
Both _pointer.Capture and CaptureGestureRecognizer should be resetting both captures and notify the current capture holder that the capture is lost

Reply:
They both do, in the Pointer class.